File: packages/hq-tools/src/tools/tts.ts
/**
 * speak — HQ Tool
 *
 * Text-to-speech via local models. Three-tier architecture:
 *   Tier 1 (default):  Kokoro-82M via kokoro-onnx (pip install kokoro-onnx soundfile)
 *                      ~600 MB RAM, ~0.3s/sentence, ONNX CPU inference
 *                      Model files: ~/.agent-hq/voice/kokoro-v1.0.int8.onnx
 *                                   ~/.agent-hq/voice/voices-v1.0.bin
 *   Tier 2 (clone):    F5-TTS MLX (pip install f5-tts-mlx)
 *                      Zero-shot voice cloning from _system/voice/reference.wav
 *   Tier 3 (fallback): macOS `say` command — 15ms, zero dependencies
 *
 * One-time setup:
 *   pip3.13 install kokoro-onnx soundfile
 *   # Model files live in ~/.agent-hq/voice/ (outside vault to avoid iCloud sync issues)
 */

import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { spawnSync } from "child_process";
import { Type } from "@sinclair/typebox";
import type { HQTool, HQContext } from "../registry.js";

// Cached availability checks (per process)
let _kokoroAvailable: boolean | null = null;
let _f5Available: boolean | null = null;

function checkPython(module: string): boolean {
  const result = spawnSync("python3.13", ["-c", `import ${module}`], {
    encoding: "utf-8",
    timeout: 5000,
  });
  return result.status === 0;
}

function isKokoroAvailable(): boolean {
  if (_kokoroAvailable !== null) return _kokoroAvailable;
  _kokoroAvailable = checkPython("kokoro_onnx") && checkPython("soundfile");
  return _kokoroAvailable;
}

function isF5Available(): boolean {
  if (_f5Available !== null) return _f5Available;
  _f5Available =
    spawnSync("f5-tts-mlx", ["--help"], { encoding: "utf-8", timeout: 5000 }).status === 0 ||
    checkPython("f5_tts_mlx");
  return _f5Available;
}

interface TTSInput {
  text: string;
  voice?: string;
  speed?: number;
  clone?: boolean;
  save?: string;
  play?: boolean;
}

const KOKORO_SCRIPT = `
import sys, os
import kokoro_onnx
import soundfile as sf

text    = sys.argv[1]
voice   = sys.argv[2] if len(sys.argv) > 2 else "af_heart"
speed   = float(sys.argv[3]) if len(sys.argv) > 3 else 1.0
out     = sys.argv[4] if len(sys.argv) > 4 else "/tmp/hq-speech.wav"
model   = sys.argv[5]
voices  = sys.argv[6]

kokoro = kokoro_onnx.Kokoro(model, voices)
samples, rate = kokoro.create(text, voice=voice, speed=speed, lang="en-us")
sf.write(out, samples, rate)
print(out)
`;

/** Resolve Kokoro model directory — prefers ~/.agent-hq/voice/, falls back to vault */
function resolveModelDir(vaultPath: string): string {
  const preferredDir = path.join(os.homedir(), ".agent-hq", "voice");
  const preferredModel = path.join(preferredDir, "kokoro-v1.0.int8.onnx");
  if (fs.existsSync(preferredModel)) return preferredDir;
  return path.join(vaultPath, "_system", "voice");
}

async function speakKokoro(
  text: string,
  voice: string,
  speed: number,
  outputPath: string,
  vaultPath: string
): Promise<boolean> {
  const modelDir   = resolveModelDir(vaultPath);
  const modelPath  = path.join(modelDir, "kokoro-v1.0.int8.onnx");
  const voicesPath = path.join(modelDir, "voices-v1.0.bin");

  if (!fs.existsSync(modelPath) || !fs.existsSync(voicesPath)) {
    console.warn(
      `[TTS/kokoro] Model files missing. Expected at ~/.agent-hq/voice/ or ${vaultPath}/_system/voice/. ` +
        "Run: mkdir -p ~/.agent-hq/voice && cp kokoro-v1.0.int8.onnx voices-v1.0.bin ~/.agent-hq/voice/"
    );
    return false;
  }

  const scriptFile = path.join(os.tmpdir(), "hq-kokoro.py");
  fs.writeFileSync(scriptFile, KOKORO_SCRIPT);

  const result = spawnSync(
    "python3.13",
    [scriptFile, text, voice, String(speed), outputPath, modelPath, voicesPath],
    { encoding: "utf-8", timeout: 30_000 }
  );

  if (result.status !== 0) {
    console.warn("[TTS/kokoro] error:", result.stderr?.slice(0, 300));
    return false;
  }
  return true;
}

async function speakF5(
  text: string,
  vaultPath: string,
  outputPath: string
): Promise<boolean> {
  const refAudio = path.join(vaultPath, "_system", "voice", "reference.wav");
  const refTranscriptPath = path.join(vaultPath, "_system", "voice", "reference-transcript.txt");

  if (!fs.existsSync(refAudio)) {
    console.warn("[TTS/f5] No reference audio at _system/voice/reference.wav");
    return false;
  }

  const refText = fs.existsSync(refTranscriptPath)
    ? fs.readFileSync(refTranscriptPath, "utf-8").trim()
    : "This is a reference audio clip.";

  const result = spawnSync(
    "f5-tts-mlx",
    ["--text", text, "--ref-audio", refAudio, "--ref-text", refText, "--output", outputPath],
    { encoding: "utf-8", timeout: 60_000 }
  );

  if (result.status !== 0) {
    console.warn("[TTS/f5] error:", result.stderr?.slice(0, 300));
    return false;
  }
  return true;
}

function speakSay(text: string): void {
  for (const v of ["Ava", "Zoe", "Samantha"]) {
    const r = spawnSync("say", ["-v", v, text], { timeout: 30_000 });
    if (r.status === 0) return;
  }
  spawnSync("say", [text], { timeout: 30_000 });
}

function playAudio(filePath: string): void {
  spawnSync("afplay", [filePath], { timeout: 60_000 });
}

export const SpeakTool: HQTool<TTSInput, string> = {
  name: "speak",
  description:
    "Generate speech audio from text using Kokoro-82M (local, free, high quality). By default only returns a [FILE:] marker so relay channels (Telegram, Discord) can send it as a voice note — the laptop speakers are NOT used unless play:true is explicitly set. Use play:true only when the user is known to be at their machine.",
  tags: ["tts", "speech", "voice", "audio", "speak", "read", "aloud", "notification", "sound"],
  schema: Type.Object({
    text: Type.String({ description: "Text to convert to speech" }),
    voice: Type.Optional(
      Type.String({
        description:
          "Kokoro voice ID. Options: af_heart (warm female, default), af_nova (clear female), am_echo (deep male), am_onyx (authoritative male), bf_emma (British female), bm_george (British male)",
      })
    ),
    speed: Type.Optional(
      Type.Number({ description: "Speech speed multiplier. Range 0.5–2.0, default 1.0" })
    ),
    clone: Type.Optional(
      Type.Boolean({
        description:
          "If true, use F5-TTS to speak in the user's own cloned voice. Requires _system/voice/reference.wav in the vault.",
      })
    ),
    play: Type.Optional(
      Type.Boolean({
        description:
          "If true, also play the audio on the laptop speakers via afplay. Default false — omit this unless the user is confirmed to be at their machine, to avoid playing audio unexpectedly.",
      })
    ),
    save: Type.Optional(
      Type.String({ description: "Optional file path to save the generated WAV audio" })
    ),
  }),

  async execute(input: TTSInput, ctx: HQContext): Promise<string> {
    const { text, voice = "af_heart", speed = 1.0, clone = false, play = false, save } = input;
    const outputPath = save ?? path.join(os.tmpdir(), `hq-speech-${Date.now()}.wav`);

    // Tier 2: Voice cloning via F5-TTS
    if (clone) {
      if (isF5Available()) {
        const ok = await speakF5(text, ctx.vaultPath, outputPath);
        if (ok) {
          if (play) playAudio(outputPath);
          return `Generated speech (F5-TTS cloned voice)\n[FILE: ${outputPath} | voice-clone.wav]`;
        }
      } else {
        console.warn("[TTS] f5-tts-mlx not installed. Run: pip install f5-tts-mlx");
      }
    }

    // Tier 1: Kokoro-82M via kokoro-onnx
    if (isKokoroAvailable()) {
      const ok = await speakKokoro(text, voice, speed, outputPath, ctx.vaultPath);
      if (ok) {
        if (play) playAudio(outputPath);
        const displayName = `voice-${voice}.wav`;
        return `Generated speech (Kokoro-82M, voice: ${voice})\n[FILE: ${outputPath} | ${displayName}]`;
      }
    } else {
      console.warn(
        "[TTS] kokoro-onnx not installed. Run: pip3.13 install kokoro-onnx soundfile"
      );
    }

    // Tier 3: macOS say — only if play explicitly requested, otherwise silent failure
    if (play) {
      speakSay(text);
      return "Spoke via macOS say (install kokoro-onnx for voice note support: pip3.13 install kokoro-onnx soundfile)";
    }
    return "TTS unavailable — install kokoro-onnx: pip3.13 install kokoro-onnx soundfile";
  },
};

File: crates/hq-tools/src/tts.rs
//! Text-to-speech tool — three-tier TTS architecture.
//!
//! Port of the TypeScript `speak` tool.
//! - Tier 1: Kokoro-82M via python3.13 + kokoro-onnx
//! - Tier 2: F5-TTS MLX for voice cloning
//! - Tier 3: macOS `say` fallback

use anyhow::Result;
use async_trait::async_trait;
use serde_json::{Value, json};
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use tokio::process::Command;

use crate::registry::HqTool;

static KOKORO_CHECKED: AtomicBool = AtomicBool::new(false);
static KOKORO_AVAILABLE: AtomicBool = AtomicBool::new(false);
static F5_CHECKED: AtomicBool = AtomicBool::new(false);
static F5_AVAILABLE: AtomicBool = AtomicBool::new(false);

async fn check_python_module(module: &str) -> bool {
    Command::new("python3.13")
        .args(["-c", &format!("import {module}")])
        .output()
        .await
        .map(|o| o.status.success())
        .unwrap_or(false)
}

async fn is_kokoro_available() -> bool {
    if KOKORO_CHECKED.load(Ordering::Relaxed) {
        return KOKORO_AVAILABLE.load(Ordering::Relaxed);
    }
    let avail = check_python_module("kokoro_onnx").await && check_python_module("soundfile").await;
    KOKORO_AVAILABLE.store(avail, Ordering::Relaxed);
    KOKORO_CHECKED.store(true, Ordering::Relaxed);
    avail
}

async fn is_f5_available() -> bool {
    if F5_CHECKED.load(Ordering::Relaxed) {
        return F5_AVAILABLE.load(Ordering::Relaxed);
    }
    let avail = Command::new("f5-tts-mlx")
        .arg("--help")
        .output()
        .await
        .map(|o| o.status.success())
        .unwrap_or(false)
        || check_python_module("f5_tts_mlx").await;
    F5_AVAILABLE.store(avail, Ordering::Relaxed);
    F5_CHECKED.store(true, Ordering::Relaxed);
    avail
}

fn resolve_model_dir(vault_path: &std::path::Path) -> PathBuf {
    let home = dirs::home_dir().unwrap_or_default();
    let preferred = home.join(".agent-hq").join("voice");
    if preferred.join("kokoro-v1.0.int8.onnx").exists() {
        preferred
    } else {
        vault_path.join("_system").join("voice")
    }
}

const KOKORO_SCRIPT: &str = r#"
import sys, os
import kokoro_onnx
import soundfile as sf

text    = sys.argv[1]
voice   = sys.argv[2] if len(sys.argv) > 2 else "af_heart"
speed   = float(sys.argv[3]) if len(sys.argv) > 3 else 1.0
out     = sys.argv[4] if len(sys.argv) > 4 else "/tmp/hq-speech.wav"
model   = sys.argv[5]
voices  = sys.argv[6]

kokoro = kokoro_onnx.Kokoro(model, voices)
samples, rate = kokoro.create(text, voice=voice, speed=speed, lang="en-us")
sf.write(out, samples, rate)
print(out)
"#;

/// Text-to-speech via local models.
pub struct SpeakTool {
    vault_path: PathBuf,
}

impl SpeakTool {
    pub fn new(vault_path: PathBuf) -> Self {
        Self { vault_path }
    }

    async fn speak_kokoro(
        &self,
        text: &str,
        voice: &str,
        speed: f64,
        output_path: &str,
    ) -> Result<bool> {
        let model_dir = resolve_model_dir(&self.vault_path);
        let model_path = model_dir.join("kokoro-v1.0.int8.onnx");
        let voices_path = model_dir.join("voices-v1.0.bin");

        if !model_path.exists() || !voices_path.exists() {
            tracing::warn!("[TTS/kokoro] Model files missing at {}", model_dir.display());
            return Ok(false);
        }

        let script_file = std::env::temp_dir().join("hq-kokoro.py");
        tokio::fs::write(&script_file, KOKORO_SCRIPT).await?;

        let output = Command::new("python3.13")
            .args([
                script_file.to_str().unwrap_or_default(),
                text,
                voice,
                &speed.to_string(),
                output_path,
                model_path.to_str().unwrap_or_default(),
                voices_path.to_str().unwrap_or_default(),
            ])
            .output()
            .await?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            tracing::warn!("[TTS/kokoro] error: {}", &stderr[..stderr.len().min(300)]);
            return Ok(false);
        }
        Ok(true)
    }

    async fn speak_f5(&self, text: &str, output_path: &str) -> Result<bool> {
        let ref_audio = self.vault_path.join("_system/voice/reference.wav");
        let ref_transcript = self.vault_path.join("_system/voice/reference-transcript.txt");

        if !ref_audio.exists() {
            tracing::warn!("[TTS/f5] No reference audio at _system/voice/reference.wav");
            return Ok(false);
        }

        let ref_text = if ref_transcript.exists() {
            tokio::fs::read_to_string(&ref_transcript)
                .await
                .unwrap_or_else(|_| "This is a reference audio clip.".to_string())
                .trim()
                .to_string()
        } else {
            "This is a reference audio clip.".to_string()
        };

        let output = Command::new("f5-tts-mlx")
            .args([
                "--text",
                text,
                "--ref-audio",
                ref_audio.to_str().unwrap_or_default(),
                "--ref-text",
                &ref_text,
                "--output",
                output_path,
            ])
            .output()
            .await?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            tracing::warn!("[TTS/f5] error: {}", &stderr[..stderr.len().min(300)]);
            return Ok(false);
        }
        Ok(true)
    }

    async fn speak_say(text: &str) {
        for voice in &["Ava", "Zoe", "Samantha"] {
            if let Ok(o) = Command::new("say").args(["-v", voice, text]).output().await {
                if o.status.success() {
                    return;
                }
            }
        }
        let _ = Command::new("say").arg(text).output().await;
    }

    async fn play_audio(path: &str) {
        let _ = Command::new("afplay").arg(path).output().await;
    }
}

#[async_trait]
impl HqTool for SpeakTool {
    fn name(&self) -> &str {
        "speak"
    }

    fn description(&self) -> &str {
        "Generate speech audio from text using Kokoro-82M (local, free, high quality). Returns a [FILE:] marker for relay channels."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "text": { "type": "string", "description": "Text to convert to speech" },
                "voice": {
                    "type": "string",
                    "description": "Kokoro voice ID: af_heart (warm female, default), af_nova, am_echo, am_onyx, bf_emma, bm_george"
                },
                "speed": { "type": "number", "description": "Speech speed 0.5-2.0 (default 1.0)" },
                "clone": { "type": "boolean", "description": "Use F5-TTS voice cloning from reference.wav" },
                "play": { "type": "boolean", "description": "Play audio on laptop speakers via afplay (default false)" },
                "save": { "type": "string", "description": "Optional file path to save the WAV" }
            },
            "required": ["text"]
        })
    }

    fn category(&self) -> &str {
        "creative"
    }

    async fn execute(&self, args: Value) -> Result<Value> {
        let text = args
            .get("text")
            .and_then(|v| v.as_str())
            .unwrap_or_default();
        let voice = args
            .get("voice")
            .and_then(|v| v.as_str())
            .unwrap_or("af_heart");
        let speed = args.get("speed").and_then(|v| v.as_f64()).unwrap_or(1.0);
        let clone = args.get("clone").and_then(|v| v.as_bool()).unwrap_or(false);
        let play = args.get("play").and_then(|v| v.as_bool()).unwrap_or(false);
        let save = args.get("save").and_then(|v| v.as_str());

        let now = chrono::Utc::now().timestamp_millis();
        let output_path = save
            .map(|s| s.to_string())
            .unwrap_or_else(|| {
                std::env::temp_dir()
                    .join(format!("hq-speech-{now}.wav"))
                    .to_string_lossy()
                    .to_string()
            });

        // Tier 2: Voice cloning
        if clone && is_f5_available().await {
            if self.speak_f5(text, &output_path).await? {
                if play {
                    Self::play_audio(&output_path).await;
                }
                return Ok(json!({
                    "message": format!(
                        "Generated speech (F5-TTS cloned voice)\n[FILE: {output_path} | voice-clone.wav]"
                    ),
                    "filePath": output_path,
                }));
            }
        }

        // Tier 1: Kokoro-82M
        if is_kokoro_available().await {
            if self.speak_kokoro(text, voice, speed, &output_path).await? {
                if play {
                    Self::play_audio(&output_path).await;
                }
                return Ok(json!({
                    "message": format!(
                        "Generated speech (Kokoro-82M, voice: {voice})\n[FILE: {output_path} | voice-{voice}.wav]"
                    ),
                    "filePath": output_path,
                }));
            }
        }

        // Tier 3: macOS say
        if play {
            Self::speak_say(text).await;
            return Ok(json!({
                "message": "Spoke via macOS say (install kokoro-onnx for voice note support)"
            }));
        }

        Ok(json!({
            "message": "TTS unavailable — install kokoro-onnx: pip3.13 install kokoro-onnx soundfile"
        }))
    }
}
